### PR TITLE
add negative prompts to validation images

### PR DESCRIPTION
**Pull request by skunkworxdark**
Add the ability to have a negative prompt in the validation image generation

I wanted to train an embedding to use in negative prompts.  So I needed the ability to produce validation images with a negative prompt to test it out while training.

### CODE REVIEW

File: src/invoke_training/_shared/stable_diffusion/validation.py
Comment (RyanJDick) [inline src/invoke_training/_shared/stable_diffusion/validation.py:23]: I *think* the `split_prompt()` logic can be replaced by a relatively simple regex pattern. Especially if we only allow `[...]` at the end of the input - which seems reasonable to me.

I'll loop back on this, unless you want to give it a shot.

**Review by RyanJDick — CHANGES_REQUESTED** [review]
Good idea! Thanks for opening this PR. There are a few changes I'd like to see before we merge:

1. The pipeline configs should be updated to accept both `validation_prompts: list[str]` and `negative_validation_prompts: list[str | None]` (with equal lengths enforced). This more explicitly defines the expected inputs for applications starting a pipeline via the API. The `[]` syntax is still great for the Gradio UI - it'll just get converted to two separate lists in order to construct the config.
2. We should write some unit tests for `split_prompt()`. It's super easy for utility functions like this, and helps tremendously with maintainability.

Let me know if you want to tackle these things. If not, I can probably get to it in a couple days.

**Comment by skunkworxdark** [comment]
> Good idea! Thanks for opening this PR. There are a few changes I'd like to see before we merge:
> 
> 1. The pipeline configs should be updated to accept both `validation_prompts: list[str]` and `negative_validation_prompts: list[str | None]` (with equal lengths enforced). This more explicitly defines the expected inputs for applications starting a pipeline via the API. The `[]` syntax is still great for the Gradio UI - it'll just get converted to two separate lists in order to construct the config.
> 2. We should write some unit tests for `split_prompt()`. It's super easy for utility functions like this, and helps tremendously with maintainability.
> 
> Let me know if you want to tackle these things. If not, I can probably get to it in a couple days.

I agree with the suggestions however, I have a busy home life over the next couple of weeks so am unlikely to get around to this any time soon. Also with your familiarity with how and where the config is implemented throughout the app, you could probably achieve this much quicker and more successfully than me. 

I assume this would mean the prompt splitting would move from the validation image generation code into the Gradio UI instead and I don't have any experience of working with that either. 

I did briefly look into doing the [] splitting with some simple regex but you get edge cases quickly unless, as you said, enforce a simple structure like only one set of [].  So I opted to use my existing code from the split_neg node in invoke.  However here is a simplified regex approach I had with and a test case, I Hope that helps a little. 

```py

def extract_strings(input_string: str) -> tuple[str, str]:
    """
    Extracts the parts of a string inside and outside of square brackets.
    
    Args:
        input_string (str): The input string to be processed.
    
    Returns:
        tuple[str, str]: A tuple containing two strings. The first string is
        everything outside the square brackets, and the second string is
        the content inside the square brackets.
        It assumes no nested brackets and doesn't work well with them.
    """
    parts = re.split(r'\[(.*?)\]', input_string)
    outside_brackets = ''.join(parts[::2])
    inside_brackets = ','.join(parts[1::2])
    return outside_brackets, inside_brackets


def test_extract_strings(self):
    # Test cases
    test_cases = [
        ("This is [inside] and [another] string", ("This is  and  string", "inside,another")),
        ("No brackets in this string", ("No brackets in this string", "")),
        ("[single bracket]", ("", "single bracket")),
    ]

    for input_string, expected_output in test_cases:
        outside, inside = extract_strings(input_string)
        self.assertEqual(outside, expected_output[0])
        self.assertEqual(inside, expected_output[1])
```

**Comment by RyanJDick** [comment]
@skunkworxdark I've addressed the things we discussed.

You'll notice that I changed the prompt syntax in the UI to be: `"positive prompt[NEG]negative prompt"`. Using a single delimiter simplifies things a bit - we don't have to worry about handling edge cases with nested brackets.

Let me know if that sounds good to you. I'll wait to merge until I get your approval.

**Comment by skunkworxdark** [comment]
At first glance, the changes looked good so I gave it a quick run and found an issue.

If you have prompts with and some without [NEG], the negative prompts of the ones without are written to the config as `- null`. This eventually resolves to a negative prompt of `None` when generating the validation images.

Is there a reason you didn't just store the empty negative prompt as an empty string and sidestep the use of None? Empty positive prompts with a negative prompt are handled this way and didn't seem to cause an issue in the brief test I did.

Just some thinking out loud thoughts:
Using the [NEG] as a delimiter is a good option as it gives a possible way for future expansion for other settings such as [CFG] or [STEPS] etc...  However, I do wonder how maintainable and easily human-read this might be long-term as it would result in multiple separate lists for each validation prompt.  Maybe the list of validation prompts should just be a single list and not multiple separate lists. It could take the form of something like a JSON representation of the required data.

**Comment by RyanJDick** [comment]
> If you have prompts with and some without [NEG], the negative prompts of the ones without are written to the config as `- null`. This eventually resolves to a negative prompt of `None` when generating the validation images.

Thanks for calling this out. This was originally done intentionally, to maintain parity with the old behaviour. But... using `negative_prompt = None` is [functionally equivalent](https://github.com/huggingface/diffusers/blob/09129842e75f9fcf1f4337a3df4dfda0c865fd7c/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L411-L412) to `negative_prompt = ""`. I've switched the type of `negative_validation_prompts` to be `list[str] | None` in [d5e618a](https://github.com/invoke-ai/invoke-training/pull/101/commits/d5e618a6ece8378c15044d73076937f7da4de9f8) - this feels slightly cleaner.


> Just some thinking out loud thoughts: Using the [NEG] as a delimiter is a good option as it gives a possible way for future expansion for other settings such as [CFG] or [STEPS] etc... However, I do wonder how maintainable and easily human-read this might be long-term as it would result in multiple separate lists for each validation prompt. Maybe the list of validation prompts should just be a single list and not multiple separate lists. It could take the form of something like a JSON representation of the required data.

I agree that this would not make for a very nice UI if we start adding more configs to the prompt lines. If we decide to do that, we'll probably want to re-design the UI for it (or make it an API-only feature).

I did consider alternative config structures (e.g. `validation_prompts: list[tuple[str, str]]`). But adding it as a separate list has the nice benefit that it does not break backwards compatibility of existing configs. The project is still young, so I'm not opposed to BC-breaking changes, but in this case the benefit doesn't feel worth it 🤷‍♂️

**Comment by RyanJDick** [comment]
The build is failing due to an issue with the CI job. All check pass locally. I'll push this through if you're happy with the current state @skunkworxdark

**Comment by skunkworxdark** [comment]
I've taken a look over the code and done a quick test run and it seems fine to me now.

**Review by RyanJDick — APPROVED** [review]

**Pull request merged**
